### PR TITLE
4.0 aesh1.0,patching,help,deployment,operators

### PR DESCRIPTION
**Pull request by jfdenise**
This PR contains the 4 major commits that are composing our move to aesh 1.0 while still keeping 100% backward compatibility.
- Patch rewrite with aesh API, better completion for ID, exclude-aged-out option
- New help support to generate synopsis and formatted help
- Deployment related commands re-implemented with new API and exposed by the "deployment" command. deploy and undeploy present still present.
- Support for | and >> operators.
- New aesh-extension grep command to grep operation/command output: e.g.: :read-resource | grep undefined

### CODE REVIEW

**Comment by bstansberry** [comment]
Don't worry about those StandardConfigsXMLValidationUnitTestCase failures.

**Comment by bstansberry** [comment]
retest this please

File: cli/src/main/java/org/jboss/as/cli/impl/CLICommandCompleter.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
```
Comment (bstansberry) [inline cli/src/main/java/org/jboss/as/cli/impl/CLICommandCompleter.java]: Please change this to the ASL copyright header. All new files in WF Core should be ASL.

File: cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java [hunk 2]
```diff
@@ -694,67 +718,73 @@ public void start() throws IOException {
...
-                        LOG.tracef("Done Executing command %s", line);
-                        loop();
-                    }
-                });
-            }, completions, preProcessors, readlineHistory, null, READLINE_FLAGS);
+                }, completions, preProcessors, readlineHistory, null, READLINE_FLAGS);
+            }
+        } catch (Exception ex) {
+            connection.write("Unexpected exception");
+            ex.printStackTrace();
```
Comment (bstansberry) [inline cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java]: Are we trying to eliminate these?

File: cli/src/main/java/org/jboss/as/cli/impl/aesh/HelpSupport.java [hunk 3]
```diff
@@ -168,4 +211,846 @@ public static String format(CommandContext ctx, BufferedReader reader) throws IO
...
+    }
+
+    public static String printHelp(CommandContext ctx, ModelNode mn,
+            OperationRequestAddress address) {
+        try {
+            // Build a ProcessedCommand from the ModelNode
+            String commandName = mn.get("operation-name").asString();
+            String desc = mn.get(Util.DESCRIPTION).asString();
+            desc += Config.getLineSeparator() + Config.getLineSeparator()
+                    + "NB: to retrieve operation full description call the following operation: "
```
Comment (bstansberry) [inline cli/src/main/java/org/jboss/as/cli/impl/aesh/HelpSupport.java]: "to retrieve the full operation description invoke" reads better

File: cli/src/main/resources/help/help.txt [hunk 4]
```diff
@@ -100,7 +100,7 @@ Usage:
  --no-operation-validation  - disable the pre validation of the operations from the CLI.
                               Operations will be only validated by the server.
 
-For a list of available commands execute
+For a list of available commands execute, once the CLI is started, execute:
```
Comment (bstansberry) [inline cli/src/main/resources/help/help.txt]: The first 'execute' should be removed.

File: cli/src/test/java/org/jboss/as/cli/parsing/test/CommandLineArgumentsTestCase.java [hunk 5]
```diff
@@ -38,22 +37,6 @@
...
-        assertTrue(args.hasProperty("--runtime-name"));
-        assertTrue(args.hasProperty("--disabled"));
-        assertTrue(args.hasProperty("--force"));
-
-        List<String> otherArgs = args.getOtherProperties();
-        assertEquals(1, otherArgs.size());
-        assertEquals("../../../../testsuite/smoke/target/deployments/test-deployment.sar", otherArgs.get(0));
-
-        assertNull(args.getOutputTarget());
-    }
```
Comment (bstansberry) [inline cli/src/test/java/org/jboss/as/cli/parsing/test/CommandLineArgumentsTestCase.java]: Does this no longer pass?

Comment (jfdenise) [reply cli/src/test/java/org/jboss/as/cli/parsing/test/CommandLineArgumentsTestCase.java]: I added it back. The operator passing is no more done by this parser but this class should still work.

File: cli/src/main/resources/help/help.txt (hunk 4 above)
Comment (jfdenise) [reply cli/src/main/resources/help/help.txt]: OK, thanks.

File: cli/src/main/java/org/jboss/as/cli/impl/aesh/HelpSupport.java (hunk 3 above)
Comment (jfdenise) [reply cli/src/main/java/org/jboss/as/cli/impl/aesh/HelpSupport.java]: Ok, thanks.

File: cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java (hunk 2 above)
Comment (jfdenise) [reply cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java]: @bstansberry , you mean the printStackTrace? This one should be printed to the user, it means a bug in CLI for which the stack should be reported.

File: cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java [hunk 6]
```diff
@@ -638,49 +666,45 @@ private void logPromptMask(String prompt, Character mask) {
...
-    @Override
     public int getTerminalHeight() {
         return getHeight();
     }
 
     private int getHeight() {
+        if (connection == null) {
+            return 40;
+        }
         if (connection instanceof TerminalConnection) {
```
Comment (stalep) [inline cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java]: why a separate checkf for TerminalConnection? cant we just call connection.size()?

File: cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java [hunk 7]
```diff
@@ -473,66 +477,97 @@ private CLITerminalConnection newConnection() throws IOException {
...
 
-    @Override
     public CommandHistory getHistory() {
         return history;
     }
 
-    @Override
     public void clearScreen() {
-        connection.stdoutHandler().accept(ANSI.CLEAR_SCREEN);
+        if (connection != null) {
```
[the remark this answers is not in the document]
Comment (jfdenise) [reply cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java]: ReadlineConsole is the backend to implement shell, so CLI shell clear calls into this method.

File: cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java (hunk 6 above)
Comment (jfdenise) [reply cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java]: Good point, removed useless cast, I guess a leftover from earlier API.

**Comment by jfdenise** [comment]
retest this please!

File: cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java (hunk 2 above)
Comment (bstansberry) [reply cli/src/main/java/org/jboss/as/cli/impl/ReadlineConsole.java]: Ok, thanks. That's fine.

**Comment by jfdenise** [comment]
This PR is now closed, the feature is covered by PR https://github.com/wildfly/wildfly-core/pull/2946

**Pull request closed without merging**